### PR TITLE
Add a data-backed `ScriptContext`

### LINKED ISSUE #6097: A `Data`-encoded `ScriptContext` (for each Plutus version) is provided in the ledger api

### PR DESCRIPTION
**Pull request by ana-pantilie**
Needs #6143
Fixes #6097

Adds a first version of a data-backed `ScriptContext`. For now, "data-backed" means that all the maps contained by the `ScriptContext` are represented as `BuiltinData`. The next step will be to make lists data-backed as well, hoping that reduces encoding/decoding time to be negligible. 

The interface provided to users is the same as the regular `ScriptContext`, the difference is that they have to import i.e. `PlutusLedgerAPI.Data.V1` instead of `PlutusLedgerAPI.V1` to use the data version. The modules which don't have anything to do with maps are shared.
Note: this might have introduced some unnecessary duplication in the code, but it doesn't seem trivial to fix it without messing up the interface. For this first version, I think this is fine but let me know if there's something we can do to improve things!

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.budget.golden [hunk 1]
```diff
@@ -0,0 +1,2 @@
+({cpu: 215441185
+| mem: 798229})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.budget.golden:2]: Non-data is:
```
({cpu: 291862695
| mem: 1092729})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.eval.golden [hunk 2]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.eval.golden:1]: This is correct.

TODO: write tests which check that the SOP and the Data versions eval to the same result

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-4.budget.golden [hunk 3]
```diff
@@ -0,0 +1,2 @@
+({cpu: 68663201
+| mem: 260597})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-4.budget.golden:2]: Non-data version:
```
({cpu: 87564279
| mem: 338905})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-4.eval.golden [hunk 4]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-4.eval.golden:1]: :+1:

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1.size.golden [hunk 5]
```diff
@@ -0,0 +1 @@
+2829
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1.size.golden:1]: Non-data version: `3183`

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-20.budget.golden [hunk 6]
```diff
@@ -0,0 +1,2 @@
+({cpu: 205140713
+| mem: 748086})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-20.budget.golden:2]: Non-data version:
```
({cpu: 281562223
| mem: 1042586})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-20.eval.golden [hunk 7]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-20.eval.golden:1]: :+1:

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-4.budget.golden [hunk 8]
```diff
@@ -0,0 +1,2 @@
+({cpu: 65614057
+| mem: 245686})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-4.budget.golden:2]: Non-data version:
```
({cpu: 84515135
| mem: 323994})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-4.eval.golden [hunk 9]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2-4.eval.golden:1]: :+1:

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2.size.golden [hunk 10]
```diff
@@ -0,0 +1 @@
+2765
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext2.size.golden:1]: Non-data version: `3119`

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityData-20.budget.golden [hunk 11]
```diff
@@ -0,0 +1,2 @@
+({cpu: 51650332
+| mem: 118202})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityData-20.budget.golden:2]: Non-data version:
```
({cpu: 67298332
| mem: 216002})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityData-20.eval.golden [hunk 12]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityData-20.eval.golden:1]: :+1:

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityOverhead-20.budget.golden [hunk 13]
```diff
@@ -0,0 +1,2 @@
+({cpu: 18656100
+| mem: 116700})
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityOverhead-20.budget.golden:2]: Non-data version:
```
({cpu: 34304100
| mem: 214500})
```

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityOverhead-20.eval.golden [hunk 14]
```diff
@@ -0,0 +1 @@
+(constr 0)
```
Comment (ana-pantilie) [inline plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContextEqualityOverhead-20.eval.golden:1]: :+1:

**Review by ana-pantilie — COMMENTED** [review]
Consistently better performance results for the Data-backed version on the `plutus-benchmark/script-contexts` tests!

File: plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.budget.golden (hunk 1 above)
Comment (Unisay) [reply plutus-benchmark/script-contexts/test/Data/9.6/checkScriptContext1-20.budget.golden:2]: 26% less cpu, 27% less of mem!

File: plutus-benchmark/script-contexts/src/PlutusBenchmark/Data/ScriptContexts.hs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+-- This example aims to *force* the decoding of the script context and then ignore it entirely.
+-- This corresponds to the unfortunate case where the decoding "wrapper" around a script forces
+-- all the decoding work to be done even if it isn't used.
+{-# INLINABLE checkScriptContext2 #-}
+checkScriptContext2 :: PlutusTx.BuiltinData -> ()
+checkScriptContext2 d =
+  let (sc :: ScriptContext) = PlutusTx.unsafeFromBuiltinData d
+  -- Just using a bang pattern was not enough to stop GHC from getting
+  -- rid of the dead binding before we even hit the plugin, this works
+  -- for now!
```
Comment (Unisay) [inline plutus-benchmark/script-contexts/src/PlutusBenchmark/Data/ScriptContexts.hs:93]: Did you consider deriving `NFData` instances and using `deepseq` to deeply evaluate the script context?

Comment (ana-pantilie) [reply plutus-benchmark/script-contexts/src/PlutusBenchmark/Data/ScriptContexts.hs:93]: This is basically all copied over from `PlutusBenchmark.ScriptContexts`, and most of the documentation in the other modules is as well copied over from the non-data versions. Which reminds me that I need properly document things, but once I refactor and remove the duplication.

File: plutus-ledger-api/src/PlutusLedgerApi/Data/V1.hs [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+So we're going to end up with multiple versions of the types and functions that we expose here, even
+internally. That means we don't lose anything by exposing all the details: we're never going to remove
+anything, we're just going to create new versions.
+-}
+
+{- | The deserialization from a serialised script into a `ScriptForEvaluation`,
+ready to be evaluated on-chain.
+Called inside phase-1 validation (i.e., deserialisation error is a phase-1 error).
+-}
+deserialiseScript ::
```
Comment (zliu41) [inline plutus-ledger-api/src/PlutusLedgerApi/Data/V1.hs:169]: I don't see why we need to redefine this. Can we not re-export the existing one?

File: plutus-ledger-api/src/PlutusLedgerApi/V2/Data/Contexts.hs [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+    , spendsOutput
+    , txSignedBy
+    , valueSpent
+    , valueProduced
+    , ownCurrencySymbol
+    ) where
+
+import GHC.Generics (Generic)
+import PlutusTx
+import PlutusTx.AssocMap hiding (filter, mapMaybe)
```
Comment (zliu41) [inline plutus-ledger-api/src/PlutusLedgerApi/V2/Data/Contexts.hs:42]: Why is it still using `AssocMap`?

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Contexts.hs [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+instance Eq ScriptPurpose where
+    {-# INLINABLE (==) #-}
+    Minting cs == Minting cs'           = cs == cs'
+    Spending ref == Spending ref'       = ref == ref'
+    Rewarding sc == Rewarding sc'       = sc == sc'
+    Certifying cert == Certifying cert' = cert == cert'
+    _ == _                              = False
+
+-- | A pending transaction. This is the view as seen by validator scripts, so some details are stripped out.
+data TxInfo = TxInfo
```
Comment (zliu41) [inline plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Contexts.hs:97]: This looks the same as the existing `TxInfo`, so why redefine it?

File: plutus-ledger-api/src/PlutusLedgerApi/V3/Data/Contexts.hs [hunk 19]
```diff
@@ -0,0 +1,697 @@
...
+findTxInByTxOutRef outRef TxInfo{txInfoInputs} =
+  PlutusTx.find
+    (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef PlutusTx.== outRef)
+    txInfoInputs
+
+{-# INLINEABLE findContinuingOutputs #-}
+
+{- | Find the indices of all the outputs that pay to the same script address we are
+currently spending from, if any.
+-}
```
Comment (effectfully) [inline plutus-ledger-api/src/PlutusLedgerApi/V3/Data/Contexts.hs:473]: `ormolu` idiocy at your service.

File: plutus-tx/src/PlutusTx/Data/AssocMap.hs [hunk 20]
```diff
@@ -77,6 +76,14 @@ instance P.UnsafeFromData (Map k a) where
   {-# INLINABLE unsafeFromBuiltinData #-}
   unsafeFromBuiltinData = Map . BI.unsafeDataAsMap
 
+instance
+  ( Pretty k
+  , Pretty a
+  , P.UnsafeFromData k
+  , P.UnsafeFromData a
+  ) => Pretty (Map k a) where
+  pretty = pretty . toList
```
Comment (effectfully) [inline plutus-tx/src/PlutusTx/Data/AssocMap.hs:85]: Why is this needed now?

**Review by effectfully — APPROVED** [review]
I really do hope we'll eventually get to cleaning it all up instead of leaving this much code duplication lying around.

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Contexts.hs (hunk 18 above)
Comment (ana-pantilie) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Contexts.hs:97]: It uses different Map and Value types, so it's not the same TxInfo.

File: plutus-tx/src/PlutusTx/Data/AssocMap.hs (hunk 20 above)
Comment (ana-pantilie) [reply plutus-tx/src/PlutusTx/Data/AssocMap.hs:85]: I believe it's for the tests.

File: plutus-ledger-api/src/PlutusLedgerApi/Data/V1.hs (hunk 16 above)
Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/Data/V1.hs:169]: Duplicating a function that uses map would be one thing, but duplicating something that doesn't use map at all makes little sense. Consider making `PlutusLedgerApi.V1` agnostic of the choice of map/list, and making `PlutusLedgerApi.V1.Term` and `PlutusLedgerApi.V1.Data` for anything specific to a particular map/list.

**Review by zliu41 — APPROVED** [review]
This looks mergeable. I think we'd need a data-encoded list as well, since using regular lists in a data-encoded ScriptContext could be very bad for efficiency.
